File: strands-ts/src/context-manager/strategies/offload/summarize.ts
/**
 * Summarize strategy — replaces oversized content with LLM-generated summaries.
 *
 * @internal
 */

import { logger } from '../../../logging/logger.js'
import { Message, TextBlock, ToolResultBlock } from '../../../types/messages.js'
import type { ContentBlock } from '../../../types/messages.js'
import type { Model } from '../../../models/model.js'
import type { LocalAgent } from '../../../types/agent.js'
import type { ContextStrategy, ContextState } from '../../types.js'
import {
  flattenMessagesToContent,
  formatSummarized,
  summarizeContent,
  toolResultToContentBlocks,
  type SummarizeConfig,
} from '../../methods/summarize.js'
import { formatStashRefs } from '../../stash.js'
import {
  BaseOffloadStrategy,
  collectRemovableWithPair,
  spliceWithPairs,
  repairAlternation,
  type OffloadConditions,
  type OffloadTarget,
} from './base.js'

export class SummarizeStrategy extends BaseOffloadStrategy {
  readonly name = 'offload:summarize'

  private readonly _config: SummarizeConfig

  constructor(target?: OffloadTarget, config?: SummarizeConfig, conditions?: OffloadConditions) {
    super(target, conditions)
    this._config = config ?? {}
  }

  when(conditions: OffloadConditions): ContextStrategy {
    return new SummarizeStrategy(this._target, this._config, conditions)
  }

  override async apply(context: ContextState): Promise<boolean> {
    if (!this._resolveModel(context.agent)) {
      logger.warn('no model available for summarization')
      return false
    }
    return super.apply(context)
  }

  protected override async _applyPerMessage(context: ContextState): Promise<boolean> {
    const model = this._resolveModel(context.agent)
    if (!model) return false

    const { messages } = context
    if (messages.length <= 1) return false

    const eligible = await this._getEligibleMessages(context)
    if (eligible.length === 0) return false

    // Expand to include paired messages so we don't orphan tool pairs
    const safeSet = new Set<Message>()
    for (const message of eligible) {
      const index = messages.indexOf(message)
      if (index === -1) continue
      for (const removable of collectRemovableWithPair(messages, index)) {
        safeSet.add(removable)
      }
    }
    const safe = messages.filter((message) => safeSet.has(message))
    if (safe.length === 0) return false

    const contentBlocks = flattenMessagesToContent(safe)
    const summary = await summarizeContent(contentBlocks, model, this._config)
    if (!summary) return false

    const totalTokens = await model.countTokens(safe)
    const summaryMessage = new Message({
      role: 'user',
      content: [new TextBlock(formatSummarized(`${safe.length} messages`, totalTokens, summary))],
    })

    const { lowestIndex } = spliceWithPairs(messages, safe)

    const insertIndex = Math.max(1, Math.min(lowestIndex, messages.length))
    messages.splice(insertIndex, 0, summaryMessage)

    repairAlternation(messages)
    logger.debug(`summarized=<${safe.length}>, tokens=<${totalTokens}> | batched summarization complete`)
    return true
  }

  protected async _replaceBlock(
    block: ContentBlock,
    tokens: number,
    message: Message,
    agent: LocalAgent,
    stashRefs: string[]
  ): Promise<ContentBlock | null> {
    const model = this._resolveModel(agent)
    if (!model) return null

    if (block instanceof ToolResultBlock) {
      const summary = await summarizeContent(toolResultToContentBlocks(block.content), model, this._config)
      if (!summary) return null

      logger.debug(`toolUseId=<${block.toolUseId}>, tokens=<${tokens}> | summarized tool result`)
      return new ToolResultBlock({
        toolUseId: block.toolUseId,
        status: block.status,
        content: [new TextBlock(`${formatSummarized('tool result', tokens, summary)}${formatStashRefs(stashRefs)}`)],
      })
    }

    if (block instanceof TextBlock) {
      const summary = await summarizeContent([new TextBlock(block.text)], model, this._config)
      if (!summary) return null

      logger.debug(`trackingId=<${message.trackingId}>, tokens=<${tokens}> | summarized text block`)
      return new TextBlock(`${formatSummarized('text block', tokens, summary)}${formatStashRefs(stashRefs)}`)
    }

    const summary = await summarizeContent([block], model, this._config)
    if (summary) {
      logger.debug(`trackingId=<${message.trackingId}>, tokens=<${tokens}> | summarized media block`)
      return new TextBlock(`${formatSummarized('media block', tokens, summary)}${formatStashRefs(stashRefs)}`)
    }

    logger.debug(`trackingId=<${message.trackingId}>, tokens=<${tokens}> | offloaded media block`)
    return new TextBlock(`[Offloaded: ~${tokens} tokens]${formatStashRefs(stashRefs)}`)
  }

  private _resolveModel(agent: LocalAgent): Model | undefined {
    return this._config.model ?? agent.model
  }
}

File: strands-py/src/strands/_context_manager/strategies/offload/summarize.py
"""Summarize strategy — replaces oversized content with LLM-generated summaries."""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING

from ....types.content import ContentBlock, Message
from ....types.tools import ToolResult, ToolResultContent
from ...methods.summarize import (
    SUMMARIZED_PREFIX,
    SummarizeConfig,
    flatten_messages_to_content,
    summarize_content,
    tool_result_to_content_blocks,
)
from ...types import ContextState
from .base import (
    BaseOffloadStrategy,
    OffloadConditions,
    OffloadTarget,
    repair_alternation,
    splice_with_pairs,
)

if TYPE_CHECKING:
    from ....agent.agent import Agent
    from ....models.model import Model

logger = logging.getLogger(__name__)


class SummarizeStrategy(BaseOffloadStrategy):
    """Summarize strategy — replaces oversized content with LLM-generated summaries."""

    @property
    def name(self) -> str:
        """Strategy name."""
        return "offload:summarize"

    def __init__(
        self,
        target: OffloadTarget | None = None,
        config: SummarizeConfig | None = None,
        conditions: OffloadConditions | None = None,
    ) -> None:
        super().__init__(target, conditions)
        self._config: SummarizeConfig = config or {}

    def when(self, **conditions: int | float) -> SummarizeStrategy:
        """Add conditions that determine when this strategy fires.

        Args:
            **conditions: Keyword arguments matching OffloadConditions fields.

        Returns:
            A new SummarizeStrategy instance with the conditions applied.
        """
        return SummarizeStrategy(self._target, self._config, OffloadConditions(**conditions))  # type: ignore[typeddict-item]

    async def apply(self, context: ContextState) -> bool:
        """Apply summarization. Returns False if no model is available.

        Args:
            context: Current context state.

        Returns:
            True if content was summarized.
        """
        if not self._resolve_model(context.agent):
            logger.warning("strategy=<%s> | no model available for summarization", self.name)
            return False
        return await super().apply(context)

    async def _apply_per_message(self, context: ContextState) -> bool:
        """Summarize oldest eligible messages into a single summary message."""
        model = self._resolve_model(context.agent)
        if not model:
            return False

        messages = context.messages
        if len(messages) <= 1:
            return False

        eligible = await self._get_eligible_messages(context)
        if not eligible:
            return False

        summarize_count = max(1, int(len(eligible) * self._removal_ratio))
        to_summarize = eligible[:summarize_count]

        content_blocks = flatten_messages_to_content(to_summarize)
        summary = await summarize_content(content_blocks, model, self._config)
        if not summary:
            return False

        total_tokens = await model.count_tokens(to_summarize)
        prefix = f"{SUMMARIZED_PREFIX} {len(to_summarize)} messages, ~{total_tokens:,} tokens]"
        summary_message = Message(
            role="user",
            content=[ContentBlock(text=f"{prefix}\n\n{summary}")],
        )

        removed, lowest_index = splice_with_pairs(messages, to_summarize)
        if removed == 0:
            return False

        insert_index = max(1, min(lowest_index, len(messages)))
        messages.insert(insert_index, summary_message)

        repair_alternation(messages)
        logger.debug("summarized=<%s>, tokens=<%s> | batched summarization complete", removed, total_tokens)
        return True

    async def _replace_block(
        self,
        block: ContentBlock,
        tokens: int,
        message: Message,
        agent: Agent,
    ) -> ContentBlock | None:
        model = self._resolve_model(agent)
        if not model:
            return None

        if "toolResult" in block:
            tool_result = block["toolResult"]
            content_blocks = tool_result_to_content_blocks(tool_result["content"])
            summary = await summarize_content(content_blocks, model, self._config)
            if not summary:
                return None

            logger.debug("tool_use_id=<%s>, tokens=<%s> | summarized tool result", tool_result["toolUseId"], tokens)
            text = f"{SUMMARIZED_PREFIX} ~{tokens:,} tokens]\n\n{summary}"
            summarized_content: list[ToolResultContent] = [{"text": text}]
            return ContentBlock(
                toolResult=ToolResult(
                    toolUseId=tool_result["toolUseId"],
                    status=tool_result["status"],
                    content=summarized_content,
                )
            )

        summary = await summarize_content([ContentBlock(text=block["text"])], model, self._config)
        if not summary:
            return None

        logger.debug("tracking_id=<%s>, tokens=<%s> | summarized text block", message.get("tracking_id"), tokens)
        return ContentBlock(text=f"{SUMMARIZED_PREFIX} ~{tokens:,} tokens]\n\n{summary}")

    def _resolve_model(self, agent: Agent) -> Model | None:
        return self._config.get("model") or agent.model
